libp2p/src/builder.rs

```diff
@@ -0,0 +1,1795 @@
+        SwarmBuilder::with_new_identity()
+            .with_tokio()
+            .with_tcp()
+            .with_tls()
+            .unwrap()
+            .with_noise()
+            .unwrap()
+            .with_dns()
+            .unwrap()
+            .with_behaviour(|_| libp2p_swarm::dummy::Behaviour)
+            .unwrap()
+            .build();
+    }
+
+    /// Showcases how to provide custom transports unknown to the libp2p crate, e.g. WebRTC.
+    #[test]
+    #[cfg(feature = "tokio")]
+    fn other_transport() -> Result<(), Box<dyn std::error::Error>> {
+        let _ = SwarmBuilder::with_new_identity()
+            .with_tokio()
+            // Closure can either return a Transport directly.
+            .with_other_transport(|_| libp2p_core::transport::dummy::DummyTransport::new())?
+            // Or a Result containing a Transport.
+            .with_other_transport(|_| {
+                if true {
+                    Ok(libp2p_core::transport::dummy::DummyTransport::new())
+                } else {
+                    Err(Box::from("test"))
+                }
+            })?
+            .with_behaviour(|_| libp2p_swarm::dummy::Behaviour)
+            .unwrap()
+            .build();
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    #[cfg(all(
+        feature = "tokio",
+        feature = "tcp",
+        feature = "tls",
+        feature = "noise",
+        feature = "dns",
+        feature = "websocket",
+    ))]
+    async fn tcp_websocket() {
+        let _ = SwarmBuilder::with_new_identity()
+            .with_tokio()
+            .with_tcp()
+            .with_tls()
+            .unwrap()
+            .with_noise()
+            .unwrap()
+            .with_websocket()
+            .with_tls()
+            .unwrap()
+            .with_noise()
+            .await
+            .unwrap()
+            .with_behaviour(|_| libp2p_swarm::dummy::Behaviour)
+            .unwrap()
+            .build();
+    }
+
+    #[tokio::test]
+    #[cfg(all(
+        feature = "tokio",
+        feature = "tcp",
+        feature = "tls",
+        feature = "noise",
+        feature = "quic",
+        feature = "dns",
+        feature = "relay",
+        feature = "websocket",
+    ))]
+    async fn all() {
+        #[derive(NetworkBehaviour)]
+        #[behaviour(prelude = "libp2p_swarm::derive_prelude")]
+        struct MyBehaviour {
+            relay: libp2p_relay::client::Behaviour,
+        }
+
+        let (builder, _bandwidth_sinks) = SwarmBuilder::with_new_identity()
+            .with_tokio()
+            .with_tcp()
+            .with_tls()
+            .unwrap()
+            .with_noise()
+            .unwrap()
+            .with_quic()
+            .with_dns()
+            .unwrap()
+            .with_relay()
+            .with_tls()
+            .unwrap()
+            .with_noise()
+            .unwrap()
+            .with_websocket()
+            .with_tls()
+            .unwrap()
+            .with_noise()
```

Comment:
Hmm, I would have expected this to work differently. Websocket is a base transport (i.e. single stream). I would have expected that I have to specify TLS and/or noise only once for these base transports.

Maybe we need to roll these things up from the other end? Like, start with the "wrapper" transports like DNS , relay and websocket and only specify what they run on last?

Reply:
Agreed that that would be the cleanest solution, but as I argue in https://github.com/libp2p/rust-libp2p/pull/4120#discussion_r1318320011 I don't think this is the easiest solution for users.

Right now each transport is build consecutively. E.g. after you called `with_tcp().with_tls().with_noise()` you are done with TCP, no later method calls will alter the TCP transport to be built.

If we follow your approach above instead, users will need to understand the difference between base transports and full transports. That includes e.g. the complexity of both TCP and relay being a base transport, while QUIC is not. I don't think we should push this complexity onto the user.

What do you think @thomaseizinger with this perspective?

Reply:
This is outdated as of our sync discussion on this. I think we should try the approach of passing sec & muxer protocols as parameters.

Reply:
> I think we should try the approach of passing sec & muxer protocols as parameters.

I don't think we can do that @thomaseizinger. See https://github.com/libp2p/rust-libp2p/pull/4120#issuecomment-1709885968